File: python/info_market_data.py
#!/usr/bin/env python3
"""
Market Data Example

Shows how to query market metadata, prices, order book, and recent trades.

The SDK handles all Info API methods automatically.

Setup:
    pip install hyperliquid-sdk

Usage:
    export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
    python info_market_data.py
"""

import os
from hyperliquid_sdk import HyperliquidSDK, HyperliquidError

ENDPOINT = os.environ.get("ENDPOINT")
if not ENDPOINT:
    print("Set ENDPOINT environment variable")
    print("Example: export ENDPOINT='https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN'")
    exit(1)

# Single SDK instance — access everything through sdk.info, sdk.core, sdk.evm, etc.
sdk = HyperliquidSDK(ENDPOINT)
info = sdk.info

print("=" * 50)
print("Market Data (Info API)")
print("=" * 50)

# Exchange metadata
print("\n1. Exchange Metadata:")
try:
    meta = info.meta()
    print(f"   Perp Markets: {len(meta.get('universe', []))}")
    for asset in meta.get("universe", [])[:5]:
        print(f"   - {asset['name']}: max leverage {asset['maxLeverage']}x")
except HyperliquidError as e:
    print(f"   (meta not available: {e.code})")

# Spot metadata
print("\n2. Spot Metadata:")
try:
    spot = info.spot_meta()
    print(f"   Spot Tokens: {len(spot.get('tokens', []))}")
except HyperliquidError as e:
    print(f"   (spot_meta not available: {e.code})")

# Exchange status
print("\n3. Exchange Status:")
try:
    status = info.exchange_status()
    print(f"   {status}")
except HyperliquidError as e:
    print(f"   (exchange_status not available: {e.code})")

# All mid prices
print("\n4. Mid Prices:")
try:
    mids = info.all_mids()
    print(f"   BTC: ${float(mids.get('BTC', 0)):,.2f}")
    print(f"   ETH: ${float(mids.get('ETH', 0)):,.2f}")
except HyperliquidError as e:
    print(f"   (allMids not available: {e.code})")

# Order book
print("\n5. Order Book (BTC):")
try:
    book = info.l2_book("BTC")
    levels = book.get("levels", [[], []])
    if levels[0] and levels[1]:
        best_bid = float(levels[0][0].get("px", 0))
        best_ask = float(levels[1][0].get("px", 0))
        spread = best_ask - best_bid
        print(f"   Best Bid: ${best_bid:,.2f}")
        print(f"   Best Ask: ${best_ask:,.2f}")
        print(f"   Spread: ${spread:.2f}")
except HyperliquidError as e:
    print(f"   (l2_book not available: {e.code})")

# Recent trades
print("\n6. Recent Trades (BTC):")
try:
    trades = info.recent_trades("BTC")
    for t in trades[:3]:
        side = "BUY" if t.get("side") == "B" else "SELL"
        print(f"   {side} {t.get('sz')} @ ${float(t.get('px', 0)):,.2f}")
except HyperliquidError as e:
    print(f"   (recent_trades not available: {e.code})")

print("\n" + "=" * 50)
print("Done!")

File: typescript/info_market_data.ts
#!/usr/bin/env npx ts-node
/**
 * Market Data Example
 *
 * Shows how to query market metadata, prices, order book, and recent trades.
 *
 * The SDK handles all Info API methods automatically.
 *
 * Setup:
 *     npm install hyperliquid-sdk
 *
 * Usage:
 *     export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
 *     npx ts-node info_market_data.ts
 */

import { HyperliquidSDK, HyperliquidError } from 'hyperliquid-sdk';

const ENDPOINT = process.env.ENDPOINT;
if (!ENDPOINT) {
  console.log("Set ENDPOINT environment variable");
  console.log("Example: export ENDPOINT='https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN'");
  process.exit(1);
}

async function main() {
  // Single SDK instance — access everything through sdk.info, sdk.core, sdk.evm, etc.
  const sdk = new HyperliquidSDK(ENDPOINT);
  const info = sdk.info;

  console.log("=".repeat(50));
  console.log("Market Data (Info API)");
  console.log("=".repeat(50));

  // Exchange metadata
  console.log("\n1. Exchange Metadata:");
  try {
    const meta = await info.meta() as Record<string, unknown>;
    const universe = (meta.universe || []) as unknown[];
    console.log(`   Perp Markets: ${universe.length}`);
    for (const asset of universe.slice(0, 5)) {
      const a = asset as Record<string, unknown>;
      console.log(`   - ${a.name}: max leverage ${a.maxLeverage}x`);
    }
  } catch (e) {
    if (e instanceof HyperliquidError) {
      console.log(`   (meta not available: ${e.code})`);
    }
  }

  // Spot metadata
  console.log("\n2. Spot Metadata:");
  try {
    const spot = await info.spotMeta() as Record<string, unknown>;
    console.log(`   Spot Tokens: ${((spot.tokens || []) as unknown[]).length}`);
  } catch (e) {
    if (e instanceof HyperliquidError) {
      console.log(`   (spotMeta not available: ${(e as HyperliquidError).code})`);
    }
  }

  // Exchange status
  console.log("\n3. Exchange Status:");
  try {
    const status = await info.exchangeStatus();
    console.log(`   ${JSON.stringify(status)}`);
  } catch (e) {
    if (e instanceof HyperliquidError) {
      console.log(`   (exchangeStatus not available: ${(e as HyperliquidError).code})`);
    }
  }

  // All mid prices
  console.log("\n4. Mid Prices:");
  try {
    const mids = await info.allMids() as Record<string, string>;
    console.log(`   BTC: $${parseFloat(mids.BTC || '0').toLocaleString()}`);
    console.log(`   ETH: $${parseFloat(mids.ETH || '0').toLocaleString()}`);
  } catch (e) {
    if (e instanceof HyperliquidError) {
      console.log(`   (allMids not available: ${(e as HyperliquidError).code})`);
    }
  }

  // Order book
  console.log("\n5. Order Book (BTC):");
  try {
    const book = await info.l2Book("BTC") as Record<string, unknown>;
    const levels = (book.levels || [[], []]) as unknown[][];
    if (levels[0]?.length && levels[1]?.length) {
      const bid0 = levels[0][0] as Record<string, unknown>;
      const ask0 = levels[1][0] as Record<string, unknown>;
      const bestBid = parseFloat(String(bid0.px || 0));
      const bestAsk = parseFloat(String(ask0.px || 0));
      const spread = bestAsk - bestBid;
      console.log(`   Best Bid: $${bestBid.toLocaleString()}`);
      console.log(`   Best Ask: $${bestAsk.toLocaleString()}`);
      console.log(`   Spread: $${spread.toFixed(2)}`);
    }
  } catch (e) {
    if (e instanceof HyperliquidError) {
      console.log(`   (l2Book not available: ${(e as HyperliquidError).code})`);
    }
  }

  // Recent trades
  console.log("\n6. Recent Trades (BTC):");
  try {
    const trades = await info.recentTrades("BTC") as unknown[];
    for (const t of trades.slice(0, 3)) {
      const trade = t as Record<string, unknown>;
      const side = trade.side === "B" ? "BUY" : "SELL";
      console.log(`   ${side} ${trade.sz} @ $${parseFloat(String(trade.px || 0)).toLocaleString()}`);
    }
  } catch (e) {
    if (e instanceof HyperliquidError) {
      console.log(`   (recentTrades not available: ${(e as HyperliquidError).code})`);
    }
  }

  console.log("\n" + "=".repeat(50));
  console.log("Done!");
}

main().catch(console.error);
